h2o-core/src/main/java/water/ExternalFrameReaderBackend.java

```diff
@@ -0,0 +1,97 @@
+package water;
+
+import water.fvec.Chunk;
+import water.fvec.ChunkUtils;
+import water.fvec.Frame;
+import water.parser.BufferedString;
+
+import java.io.IOException;
+import java.nio.channels.SocketChannel;
+import java.util.UUID;
+
+
+import static water.ExternalFrameHandler.*;
+import static water.ExternalFrameUtils.*;
+
+/**
+ * This class contains methods used on the h2o backend to read data as from H2O Frame
+ */
+class ExternalFrameReaderBackend {
+
+    /**
+     * Internal method use on the h2o backend side to handle reading from the chunk from non-h2o environment
+     * @param channel socket channel originating from non-h2o node
+     * @param initAb {@link AutoBuffer} containing information necessary for preparing backend for reading
+     */
+    static void handleReadingFromChunk(SocketChannel channel, AutoBuffer initAb) throws IOException {
+        String frameKey = initAb.getStr();
+        int chunkIdx = initAb.getInt();
+        byte[] expectedTypes = initAb.getA1();
+        int[] selectedColumnIndices = initAb.getA4();
+        assert selectedColumnIndices != null;
+        Frame fr = DKV.getGet(frameKey);
+        Chunk[] chunks = ChunkUtils.getChunks(fr, chunkIdx);
+        if(expectedTypes == null){
+            expectedTypes = prepareExpectedTypes(fr.vecs());
+        }
+
+        AutoBuffer ab = new AutoBuffer().flipForReading().clearForWriting(H2O.MAX_PRIORITY);
+        ab.putInt(chunks[0]._len); // write num of rows
+        writeToChannel(ab, channel);
+
+        // buffer string to be reused for strings to avoid multiple allocation in the loop
+        BufferedString valStr = new BufferedString();
+        for (int rowIdx = 0; rowIdx < chunks[0]._len; rowIdx++) { // for each row
+            for(int cidx: selectedColumnIndices){ // go through the chunks
+                if (chunks[cidx].isNA(rowIdx)) {
+                    ExternalFrameUtils.sendNA(ab, channel,expectedTypes[cidx]);
+                } else {
+                    final Chunk chnk = chunks[cidx];
+                    switch (expectedTypes[cidx]) {
```

Comment:
Yup - However I would proceed with this code and consider this as an addition in the future, so we can finally put this into master, this would be bigger change. 

Maybe later we keep both variants and configure it using some configuration property - either batch mode or on-the-fly mode